### PR TITLE
feat: backportbot 2.0.0

### PR DESCRIPTION
**Pull request by skjnldsv**

### CODE REVIEW

File: src/queue.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import PQueue from 'p-queue'
+import { Task } from './constants'
+import { backport } from './backport'
+
+let queue: PQueue
+
+export const addToQueue = (task: Task): Promise<void> => {
+	if (!queue) {
+		queue = new PQueue({ concurrency: 1 })
```
Comment (skjnldsv) [inline src/queue.ts:9]: For now, it's really fast and I had issues with concurrency. Let's keep it to 1

File: src/nextcloudUtils.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export const getMilestoneFromBase = (branch: string, milestones: Milestone[]): Milestone => {
+	// Extract the version from the branch name, e.g. stable21
+	const version = branch.match(/stable(\d{2})/i)?.[1]
+	if (!version) {
+		throw new Error(`Could not extract version from branch ${branch}`)
+	}
+	const selection = milestones
+		.filter(milestone => milestone.title.startsWith('Nextcloud ' + version))
```
Comment (nickvergessen) [inline src/nextcloudUtils.ts:33]: Could we return to https://github.com/nextcloud/backportbot/pull/388/files#diff-b7b030aed4ec199a8bc4801602fe6c4a72868e111f692b08d08278bb1946ce3cL15-L16
So it works for Talk's milestones again?
https://github.com/nextcloud/spreed/milestones

```suggestion
		.filter(milestone => milestone.title.contains(version))
```

That would work a bit better with "decoupled" apps that have other majors/minors than server.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+				info(`\nReceived full backport request to \`${branch}\``)
+				info(`├ Fetching commits from PR ${htmlUrl}...`)
+				commits = await getCommitsForPR(authOctokit, owner, repo, prNumber)
+			} else {
+				info(`\nReceived partial backport request to \`${branch}\``)
+			}
+
+			// PR info
+			if (isForcedRequest && isMerged) {
+				info(`├ PR is merged, but force flag is present, starting backport right away`)
```
Comment (nickvergessen) [inline src/index.ts:206]: do you mean to log this?
```suggestion
			if (isForcedRequest && !isMerged) {
				info(`├ PR is not merged, but force flag is present, starting backport right away`)
```

File: src/nextcloudUtils.ts (hunk 2 above)
Comment (skjnldsv) [reply src/nextcloudUtils.ts:33]: I'll use the old way, sure :)

Comment (skjnldsv) [reply src/nextcloudUtils.ts]: Done

File: src/constants.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+export const SERVE_PORT = process.env.SERVE_PORT || 3000
+export const SERVE_HOST = process.env.SERVE_HOST || '0.0.0.0'
+
+// Path variables
+export const ROOT_DIR = resolve(__dirname + '/../')
+export const CACHE_DIRNAME = 'cache' // relative to the root dir
+export const WORK_DIRNAME = 'work' // relative to the root dir
+
+// Nextcloud variables
+export const ALLOWED_ORGS = ['nextcloud', 'nextcloud-libraries', 'skjnldsv']
```
Comment (vitormattos) [inline src/constants.ts:13]: Would be nice move this to environment and not be a constant to be possible use this bot by other organizations.

File: src/nextcloudUtils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import { describe, expect, test } from 'vitest'
+import { Milestone } from '@octokit/webhooks-types'
+import { getMilestoneFromBase } from './nextcloudUtils'
+
+describe('Match branch branch to milestone', () => {
+	const branches = [
+		'stable28',
```
Comment (nickvergessen) [inline src/nextcloudUtils.spec.ts:7]:
```suggestion
		'stable28',
		'stable26',
```

File: src/nextcloudUtils.spec.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		'stable21',
+		'stable1',
+		'stable-3.11',
+		'stable-4.0.0',
+	]
+
+	const milestones = [
+		'Nextcloud 21.1.0',
```
Comment (nickvergessen) [inline src/nextcloudUtils.spec.ts:15]:
```suggestion
		'Nextcloud 26.0.10',
		'Nextcloud 26.0.9',
		'Nextcloud 21.1.0',
```

File: src/nextcloudUtils.spec.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		'Nextcloud 1.1.1',
+		'💞 Next Major (29)',
+		'💙 Next Patch (28)',
+		'💔 Backlog',
+		'3.11.1',
+		'3.10.4',
+	].map(title => ({ title })) as Milestone[]
+
+	const expectedMilestones = [
+		'💙 Next Patch (28)',
```
Comment (nickvergessen) [inline src/nextcloudUtils.spec.ts:27]:
```suggestion
		'💙 Next Patch (28)',
		'Nextcloud 26.0.9',
```

**Review by nickvergessen — COMMENTED** [review]
A good sample to add

File: src/constants.ts (hunk 4 above)
Comment (skjnldsv) [reply src/constants.ts:13]: Anyone is free to fork and run their own :)

Comment (vitormattos) [reply src/constants.ts:13]: But, maybe now is a good moment to implement this as environment, like the other points at this file.

Comment (vitormattos) [reply src/constants.ts:13]: When we have `skjnldsv` it is an indicator that being a configuration environment is a good way to go

File: README.md [hunk 8]
```diff
@@ -1,41 +1,37 @@
...
-## Setup
-
-```sh
-# Install dependencies
-npm install
-
-# Run the bot
-npm start
-```
+The Nextcloud Backportbot is a GitHub bot designed to streamline the process of backporting pull requests or specific commits to older branches in the Nextcloud repository.
```
Comment (vitormattos) [inline README.md:3]: Would be good maintain this bot not only able to run at Nextcloud organization but will be very nice be possible other app developers to have an instance of this bot running in other places to be used by repositories of apps outside Nexcloud organization.

Signaling that this project can be used outside of the Nextcloud organization also helps attract more contributors to this project.

```suggestion
The Nextcloud Backportbot is a GitHub bot designed to streamline the process of backporting pull requests or specific commits to older branches.
```

File: README.md [hunk 9]
```diff
@@ -1,41 +1,37 @@
-# backport
+# Nextcloud Backportbot
```
Comment (vitormattos) [inline README.md:1]: Would be good maintain this bot not only able to run at Nextcloud organization but will be very nice be possible other app developers to have an instance of this bot running in other places to be used by repositories of apps outside Nexcloud organization.

Signaling that this project can be used outside of the Nextcloud organization also helps attract more contributors to this project.

```suggestion
# Backportbot
```

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**